sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go

```diff
@@ -58,6 +58,10 @@ func GetBeamJar(gradleTarget, version string) (string, error) {
 }
 
 func (j *jarGetter) getJar(gradleTarget, version string) (string, error) {
+	if strings.Contains(version, ".dev") {
```

Comment:
While I support having a clearer error message for users, it feels a bit limiting to just disallow any container with ".dev" outright. What about approaching this as letting the code fail naturally, and if it fails and we detect that the user was attempting to retrieve a .dev container, then we add this message.

You'd need to run this code with a .dev container to see where exactly it fails (probably on http.Get if I had to guess). But then all you have to do is move this if statement to that error block, and if it's true then wrapping the error with a [SetTopLevelMsg](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/internal/errors/errors.go#L90) function should do it. (I originally wrote that function for this exact purpose; adding messages directly to users to tell them how to fix a specific error).

That changes this behavior from "disallowing something" to "detecting a common error case".

Reply:
It's worth noting that the Python SDK also routes .dev versioning through gradle ([code reference here](https://github.com/apache/beam/blob/8e217ea0d1f383ef5033ef507b14d01edf9c67e6/sdks/python/apache_beam/utils/subprocess_server.py#L233)) I could move the check to after the local cache check to fully emulate the Python if we wanted to though.

Reply:
It seems reasonable to leave as-is if it's functionally equivalent to the Python behavior. And checking the local cache first should mean that users will be able to use local jars they've already built pretty easily. So I'll approve it now.